pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -422,67 +440,78 @@ func GetPipelineConditionStatus(pr *v1beta1.PipelineRun, state PipelineRunState,
-				Message: fmt.Sprintf("TaskRun %s has failed", rprt.TaskRun.Name),
-			}
-		}
-	}
-
 	allTasks := []string{}
-	successOrSkipTasks := []string{}
+	withStatusTasks := []string{}
 	skipTasks := int(0)
+	failedTasks := int(0)
+	cancelledTasks := int(0)
+	reason := ReasonSucceeded
 
 	// Check to see if all tasks are success or skipped
+	//
+	// The completion reason is also calculated here, but it will only be used
+	// if all tasks are completed.
+	//
+	// The pipeline run completion reason is set from the taskrun completion reason
+	// according to the following logic:
+	//
+	// - All successful: ReasonSucceeded
+	// - Some successful, some skipped: ReasonCompleted
+	// - Some cancelled, none failed: ReasonCancelled
+	// - At least one failed: ReasonFailed
 	for _, rprt := range state {
 		allTasks = append(allTasks, rprt.PipelineTask.Name)
-		if rprt.IsSuccessful() {
-			successOrSkipTasks = append(successOrSkipTasks, rprt.PipelineTask.Name)
-		}
-		if isSkipped(rprt, state.ToMap(), dag) {
+		switch {
+		case rprt.IsSuccessful():
+			withStatusTasks = append(withStatusTasks, rprt.PipelineTask.Name)
+		case isSkipped(rprt, state.ToMap(), dag):
 			skipTasks++
-			successOrSkipTasks = append(successOrSkipTasks, rprt.PipelineTask.Name)
+			withStatusTasks = append(withStatusTasks, rprt.PipelineTask.Name)
+			// At least one is skipped and no failure yet, mark as completed
+			if reason == ReasonSucceeded {
+				reason = ReasonCompleted
+			}
+		case rprt.IsCancelled():
+			cancelledTasks++
+			withStatusTasks = append(withStatusTasks, rprt.PipelineTask.Name)
+			if reason != ReasonFailed {
+				reason = ReasonCancelled
+			}
+		case rprt.IsFailure():
+			withStatusTasks = append(withStatusTasks, rprt.PipelineTask.Name)
+			failedTasks++
+			reason = ReasonFailed
 		}
 	}
 
-	if reflect.DeepEqual(allTasks, successOrSkipTasks) {
-		logger.Infof("All TaskRuns have finished for PipelineRun %s so it has finished", pr.Name)
-		reason := ReasonSucceeded
-		if skipTasks != 0 {
-			reason = ReasonCompleted
+	if reflect.DeepEqual(allTasks, withStatusTasks) {
+		status := corev1.ConditionTrue
+		if failedTasks > 0 || cancelledTasks > 0 {
+			status = corev1.ConditionFalse
 		}
-
+		logger.Infof("All TaskRuns have finished for PipelineRun %s so it has finished", pr.Name)
 		return &apis.Condition{
-			Type:    apis.ConditionSucceeded,
-			Status:  corev1.ConditionTrue,
-			Reason:  reason,
-			Message: fmt.Sprintf("Tasks Completed: %d, Skipped: %d", len(successOrSkipTasks)-skipTasks, skipTasks),
+			Type:   apis.ConditionSucceeded,
+			Status: status,
+			Reason: reason,
+			Message: fmt.Sprintf("Tasks Completed: %d (Failed: %d, Cancelled %d), Skipped: %d",
+				len(allTasks)-skipTasks, failedTasks, cancelledTasks, skipTasks),
 		}
 	}
 
-	// Hasn't timed out; no taskrun failed yet; and not all tasks have finished....
+	// Hasn't timed out; not all tasks have finished....
 	// Must keep running then....
+	if failedTasks > 0 || cancelledTasks > 0 {
```

Comment:
Isn't this point reached if there are dependent tasks that haven't been (and won't be) scheduled?  I don't see anything distinguishing running from not yet running.

Reply:
Uhm, when I wrote this code I understood that isSkipped would mark descendants of failed tasks as skipped, but it doesn't, so I may need to revisit that, and add more unit tests because that should be caught by testing if it's not working as it should.

I thought the case you mentioned was covered but I'll check that again

Reply:
Uhm, when I wrote this code I understood that isSkipped would mark descendants of failed tasks as skipped, but it doesn't, so I may need to revisit that, and add more unit tests because that should be caught by testing if it's not working as it should.

I thought the case you mentioned was covered but I'll check that again

Reply:
@afrittoli havent looked all the changes yet but, skipped tasks are very tricky, please see my comment [here](https://github.com/tektoncd/pipeline/pull/2661#discussion_r428827199)

Reply:
Thanks for catching this issue!

I've added now logic to distinguish the running and non running cases.
When the pipeline is in stopping mode, taskrun that have not been started yet will be counted as skipped. So if there is any task that is running the pipeline stays in stopping, else it switches to failed.

I've added more test cases for this too.

Reply:
👍